### PR TITLE
fix runtime for dogecoin on Alpine Linux (musl libc)

### PR DESCRIPTION
**Pull request by chey**
`dogecoind` built with musl libc will die unexpectedly giving a `Segmentation fault (core dumped)`.

This patch makes it so dogecoin will run successfully on Alpine Linux using the standard `musl` libc.

### CODE REVIEW

**Comment by chey** [comment]
Here is some more detail on how I pinned this down.

After seeing the Segfault I installed the `gdb` and here is the output:
```
/dogecoin/src/crypto # gdb --args dogecoind -disablewallet -printtoconsole
GNU gdb (GDB) 8.2
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-alpine-linux-musl".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from dogecoind...done.
```
```
(gdb) run
Starting program: /usr/local/bin/dogecoind -disablewallet -printtoconsole
warning: Error disabling address space randomization: Operation not permitted
2021-06-02 23:10:37

2021-06-02 23:10:37 Dogecoin version v1.14.4.0-unk
2021-06-02 23:10:37 InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
2021-06-02 23:10:37 Assuming ancestors of block 77e3f4a4bcb4a2c15e8015525e3d15b466f6c022f6ca82698f329edef7d9777e have valid signatures.
2021-06-02 23:10:37 Default data directory /root/.dogecoin
2021-06-02 23:10:37 Using data directory /root/.dogecoin
2021-06-02 23:10:37 Using config file /root/.dogecoin/dogecoin.conf
2021-06-02 23:10:37 Using at most 125 automatic connections (1048576 file descriptors available)
2021-06-02 23:10:37 Using 32 MiB out of 32 requested for signature cache, able to store 1048576 elements
2021-06-02 23:10:37 Using 4 threads for script verification
[New LWP 70]
[New LWP 71]
[New LWP 72]
[New LWP 73]
2021-06-02 23:10:37 libevent: getaddrinfo: nodename nor servname provided, or not known
2021-06-02 23:10:37 Binding RPC on address ::1 port 22555 failed.
2021-06-02 23:10:37 HTTP: creating work queue of depth 16
2021-06-02 23:10:37 No rpcpassword set - using random cookie authentication
2021-06-02 23:10:37 Generated RPC authentication cookie /root/.dogecoin/.cookie
2021-06-02 23:10:37 HTTP: starting 4 worker threads
2021-06-02 23:10:37 scheduler thread start
[New LWP 74]
[New LWP 75]
[New LWP 76]
[New LWP 77]
[New LWP 78]
2021-06-02 23:10:37 Bound to [::]:22556
2021-06-02 23:10:37 Bound to 0.0.0.0:22556
2021-06-02 23:10:37 Cache configuration:
2021-06-02 23:10:37 * Using 2.0MiB for block index database
2021-06-02 23:10:37 * Using 8.0MiB for chain state database
2021-06-02 23:10:37 * Using 440.0MiB for in-memory UTXO set (plus up to 286.1MiB of unused mempool space)
2021-06-02 23:10:37 init message: Loading block index...
2021-06-02 23:10:37 Opening LevelDB in /root/.dogecoin/blocks/index
2021-06-02 23:10:37 Opened LevelDB successfully
2021-06-02 23:10:37 Using obfuscation key for /root/.dogecoin/blocks/index: 0000000000000000
2021-06-02 23:10:37 Opening LevelDB in /root/.dogecoin/chainstate
2021-06-02 23:10:37 Opened LevelDB successfully
2021-06-02 23:10:37 Using obfuscation key for /root/.dogecoin/chainstate: 2f202083e75610ed
2021-06-02 23:10:37 LoadBlockIndexDB: last block file = 0
2021-06-02 23:10:37 LoadBlockIndexDB: last block file info: CBlockFileInfo(blocks=2, size=464, heights=0...0, time=2013-12-06...2013-12-06)
2021-06-02 23:10:37 Checking all blk files are present...
2021-06-02 23:10:37 LoadBlockIndexDB: transaction index disabled
2021-06-02 23:10:37 Initializing databases...
2021-06-02 23:10:37 init message: Verifying blocks...
2021-06-02 23:10:37  block index               7ms
2021-06-02 23:10:37 Wallet disabled!
[New LWP 79]

Thread 11 "dogecoin-loadbl" received signal SIGSEGV, Segmentation fault.
[Switching to LWP 79]
0x00005584cb138dcd in scrypt_1024_1_1_256 (input=input@entry=0x5584ccffbd70 "\001",
    output=output@entry=0x7fbe95d91ae0 "") at ../../src/crypto/scrypt.cpp:258
warning: Source file is more recent than executable.
258	    scrypt_1024_1_1_256_sp(input, output, scratchpad);
```

```
(gdb) bt
#0  0x00005584cb138dcd in scrypt_1024_1_1_256 (input=input@entry=0x5584ccffbd70 "\001",
    output=output@entry=0x7fbe95d91ae0 "") at ../../src/crypto/scrypt.cpp:258
#1  0x00005584cb061ee3 in CPureBlockHeader::GetPoWHash (this=this@entry=0x5584ccffbd70)
    at ../../src/primitives/pureheader.cpp:28
#2  0x00005584caffcdc4 in CheckAuxPowProofOfWork (block=..., params=...) at ../../src/dogecoin.cpp:106
#3  0x00005584cafc01d5 in ReadBlockOrHeader<CBlock> (block=..., pos=..., consensusParams=..., fCheckPOW=<optimized out>)
    at /usr/include/c++/8.3.0/bits/shared_ptr_base.h:1167
#4  0x00005584cafc0310 in ReadBlockOrHeader<CBlock> (block=..., pindex=0x5584ccffbf60, consensusParams=...,
    fCheckPOW=<optimized out>) at ../../src/chain.h:254
#5  0x00005584cafc5806 in ReadBlockFromDisk (fCheckPOW=true, consensusParams=..., pindex=<optimized out>, block=...)
    at ../../src/validation.cpp:1192
#6  ConnectTip (connectTrace=..., pblock=..., pindexNew=<optimized out>, chainparams=..., state=...)
    at ../../src/validation.cpp:2280
#7  ActivateBestChainStep (state=..., chainparams=..., pindexMostWork=0x5584ccffbf60, pblock=...,
    fInvalidFound=@0x7fbe95d91f90: false, connectTrace=...) at ../../src/validation.cpp:2429
#8  0x00005584cafc6450 in ActivateBestChain (state=..., chainparams=..., pblock=...) at ../../src/uint256.h:47
#9  0x00005584cae8b43f in ThreadImport (vImportFiles=...) at /usr/include/c++/8.3.0/bits/shared_ptr_base.h:614
#10 0x00005584caea2fd4 in boost::_bi::list1<boost::_bi::value<std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> > > >::operator()<void (*)(std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> >), boost::_bi::list0> (f=<optimized out>, a=<synthetic pointer>..., this=<optimized out>)
    at /usr/include/boost/bind/bind.hpp:1291
#11 boost::_bi::bind_t<void, void (*)(std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> >), boost::_bi::list1<boost::_bi::value<std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> > > > >::operator() (this=<optimized out>) at /usr/include/boost/bind/bind.hpp:1294
#12 boost::detail::thread_data<boost::_bi::bind_t<void, void (*)(std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> >), boost::_bi::list1<boost::_bi::value<std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> > > > > >::run (this=<optimized out>) at /usr/include/boost/thread/detail/thread.hpp:117
#13 0x00007fbe98632065 in ?? () from /usr/lib/libboost_thread-mt.so.1.67.0
#14 0x00007fbe987476a2 in ?? () from /lib/ld-musl-x86_64.so.1
#15 0x0000000000000000 in ?? ()
```

**Comment by chey** [comment]
@haxpor ^ regarding your question

**Comment by patricklodder** [comment]
@chey Do you have a Dockerfile / alpine image link that I can use to reproduce?

**Comment by chey** [comment]
> @chey Do you have a Dockerfile / alpine image link that I can use to reproduce?

Here is the one I'm using.

https://github.com/chey/dogecoin/blob/alpine-dockerfile/alpine.Dockerfile

**Comment by nooperation** [comment]
Related:
https://github.com/litecoin-project/litecoin/issues/407
https://wiki.musl-libc.org/functional-differences-from-glibc.html#Thread_stack_size

Tldr:
Scratchpad requires 131135 bytes on the stack, but musl only allows for a stack size of 80k or 128k depending on version. Allocating scratchpad in the data segment via static local resolves it. This change is pretty sketchy if you have multiple threads hitting it at the same time, but I'm not sure if this would be the case with this codebase

**Comment by patricklodder** [comment]
@nooperation Thank you!

I was trying to reproduce by building with [a Dockerfile derived from Author's linked one](https://gist.github.com/patricklodder/1238774314a0916f70f6b8f74599404d#file-dockerfile-alpine-withtests) and running `src/test/test_dogecoin --run_test=scrypt_tests` but this didn't succeed to reproduce the issue using the unit tests, even though a similar scratchpad function is used at [`scrypt_tests.cpp#L21`](https://github.com/dogecoin/dogecoin/blob/e34197907f32f3fca371bfc490ef6e521723a070/src/test/scrypt_tests.cpp#L21)

Perhaps we could build with an a larger thread stack size instead, per your [linked documentation](https://wiki.musl-libc.org/functional-differences-from-glibc.html#Thread_stack_size)?

**Comment by chey** [comment]
> Perhaps we could build with an a larger thread stack size instead, per your [linked documentation](https://wiki.musl-libc.org/functional-differences-from-glibc.html#Thread_stack_size)?

I believe what the documentation is saying is that one could compile musl-libc itself with a larger stack size and is not referring to "the app", `dogecoind` in this case.

**Comment by nooperation** [comment]
I'm uncertain as to what the correct solution to this one is because I'm fairly unfamiliar with building on these systems. The docs make it seem like you can specific a thread stack size, but I haven't been able to reproduce that successfully. I mainly just don't like these specific changes because they may have very unintended impacts relating to crypto stuff (mainly the scratchpad that can be modified by multiple threads at the same exact time). The lowest impact solution might just be to state that it's not supported. 

As for reproduction? It might be that you're not running into the issue in the tests because they're being run on the main thread. You may need to create a thread before calling scrypt_1024_1_1_256 (or CBlockHeader::GetPoWHash()). Something simple like:

```
#include <iostream>
#include <thread>

static const int SCRYPT_SCRATCHPAD_SIZE = 131072 + 63;

void scrypt_1024_1_1_256(const char *input, char *output)
{
    char scratchpad[SCRYPT_SCRATCHPAD_SIZE];
    printf("Hello world");
}

int main()
{
    std::thread([] {
            char output[32] = {};
            scrypt_1024_1_1_256("Hello", output);
    }).join();

    return 0;
}
```

**Comment by rnicoll** [comment]
Verified with and without on Alpine Linux 3.14.0 on AWS, and looks good.

**Review by rnicoll — APPROVED** [review]
Tested ACK. Also basically this just manually does what the libc should be handling automatically (ensuring it doesn't frag the memory), so worst case it's slightly more code for zero impact.

**Comment by rnicoll** [comment]
Realising.... is this going to be thread safe? As in, do we have cases where it hashes multiple blocks from different threads?

**Review by rnicoll — CHANGES_REQUESTED** [review]
Gyah - sorry, yeah, this wouldn't be thread safe and validation can happen on multiple threads (even if it generally doesn't).

We can put a lock around this, but that then slows every other platform.

Anyone got a good solution for something like:

```
#ifdef LIB_MUSL
[acquire lock]
[use static implementation]
#else
[use stack implementation]
#endif
```

Needs some way of detecting musl at configure time, though

**Comment by chromatic** [comment]
As far as I can tell, [there's no MUSL macro by design](https://wiki.musl-libc.org/faq.html#Q:_Why_is_there_no_%3Ccode%3E__MUSL__%3C/code%3E_macro?). Probably means a double check for Linux and *not* `__gnu_linux__`.

Alternately, a recent enough version of GCC might [define OPTION_MUSL](https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/config/linux.h;h=4e1db60fcedaa23499aa9dc79577397c17e91e4e;hb=HEAD#l30) if I read this right.

**Comment by Bills1Drive** [comment]
Will this help? 
https://alpinelinux.org/posts/Alpine-Linux-has-switched-to-musl-libc.html

**Comment by Bills1Drive** [comment]
The important step is the first one.
1.
Edit /etc/apk/repositories and make sure that you don’t mix edge and any stable branch (eg v2.x)

**Comment by nooperation** [comment]
One alternative is to look at `thread_local` if it's supported here. It would give each thread its own scratchpad and it would also take it off of the stack, resolving the original issue

```
void scrypt_1024_1_1_256(const char *input, char *output)
{
    thread_local char scratchpad[SCRYPT_SCRATCHPAD_SIZE];
    scrypt_1024_1_1_256_sp(input, output, scratchpad);
}
```

**Comment by chey** [comment]
> One alternative is to look at `thread_local` if it's supported here. It would give each thread its own scratchpad and it would also take it off of the stack, resolving the original issue
> 
> ```
> void scrypt_1024_1_1_256(const char *input, char *output)
> {
>     thread_local char scratchpad[SCRYPT_SCRATCHPAD_SIZE];
>     scrypt_1024_1_1_256_sp(input, output, scratchpad);
> }
> ```

I've tested this myself as well and it works.

All my tests pass and so do the 253 tests that are compiled into `test_dogecoin`.

**Comment by kaniini** [comment]
Using `thread_local` here is the correct fix, and it should be done regardless of whether or not you're building on musl, because its still broken on glibc, its just that glibc gives threads 8MB of stack by default.

**Comment by patricklodder** [comment]
> All my tests pass and so do the 253 tests that are compiled into test_dogecoin.

@chey could you please update your PR to use `thread_local` then? It'll allow us to move this forward.

Thanks everyone for the help! :heart:

**Review by patricklodder — APPROVED** [review]
Tested ACK. 

- Fixes the problem when compiling with musl ✅ 
- Able to sync testnet on musl ✅
- No regressions on testnet and mainnet full sync w/ x64_86-pc-linux-gnu ✅

[Mined a testnet block w/ glibc](https://chain.so/block/DOGETEST/fb47b2451f30c4d15aebca36ffc66c41d55c2f33684543076451c94ffe335217)

[Mined a testnet block w/ musl](https://chain.so/block/DOGETEST/f0ef5edaa00b9e982b10cbdf8ef46b8535040552a9548ccb11822f6164fcbf0f)

All good!

**Review by rnicoll — APPROVED** [review]
utACK
